src/Linux/Containers/Impl/Docker/LinuxDockerService.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.IO;
+using Microsoft.Common.Core.OS;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+using System.Threading;
+using Newtonsoft.Json.Linq;
+using System.IO;
+
+namespace Microsoft.R.Containers.Docker {
+    public class LinuxDockerService : IContainerService {
+        const string DockerCePackageName = "docker-ce";
+        const string DockerEePackageName = "docker-ee";
+        const string DockerProcessName = "dockerd";
+
+        private readonly LocalDocker _docker;
+        private readonly IFileSystem _fs;
+        private readonly IProcessServices _ps;
+        private readonly Regex _containerIdMatcher64 = new Regex("[0-9a-f]{64}", RegexOptions.IgnoreCase);
+        private readonly Regex _containerIdMatcher12 = new Regex("[0-9a-f]{12}", RegexOptions.IgnoreCase);
+
+        public LinuxDockerService(IFileSystem fs, IProcessServices ps) {
+            _fs = fs;
+            _ps = ps;
+            _docker = GetLocalDocker();
+        }
+
+        public ContainerServiceStatus GetServiceStatus() {
+            ContainerServiceStatus status = new ContainerServiceStatus();
+            var proc = GetDockerProcess();
+            if (proc.HasExited) {
+                status.IsServiceAvailable = false;
+                status.StatusType = ContainerServiceStatusType.Error;
+                status.StatusMessage = Resources.Error_ServiceNotAvailable;
+            } else {
+                status.IsServiceAvailable = true;
+                status.StatusType = ContainerServiceStatusType.Information;
+                status.StatusMessage = Resources.Info_ServiceAvailable;
+            }
+            return status;
+        }
+
+        internal static Process GetDockerProcess() {
+            var processes = Process.GetProcessesByName(DockerProcessName);
+            return (processes.Count() >= 1) ? processes.First() : null;
+        }
+
+        public async Task<IContainer> GetContainerAsync(string containerId, CancellationToken ct) {
+            string output = await _docker.ListContainersAsync(_ps, true, ct);
+            var ids = output.Split(new string[] { "\n", "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (string id in ids) {
```

Comment:
Its probable possible to even build `async Select` query, @AlexanderSher is pro there.